src/lib.rs

```diff
@@ -210,52 +198,111 @@ impl TryFrom<u8> for ControllerDataPackageType {
-pub fn parse<#[cfg(feature = "heapless")] const MAX_RESULTS: usize>(
-    input: &[u8],
-) -> ParseResult<MAX_RESULTS> {
-    /// Simple state machine for the parser, represents whether the parser is seeking a start or has found it.
-    enum ParserState {
-        SeekStart,
-        ParseCommand,
+/// Parse the input string for commands.
+///
+/// Null bytes (`b"\x00"`) will be skipped completely, unparseable content will return `Some(Err(ProtocolParseError))`
+/// but will not fail the parsing completely (i.e. you can continue to get the next entry until you reach the end of the input).
+///
+/// ## Example
+/// ```
+/// # use adafruit_bluefruit_protocol::button_event::{Button, ButtonParseError, ButtonState};
+/// # use adafruit_bluefruit_protocol::ControllerEvent::ButtonEvent;
+/// # use adafruit_bluefruit_protocol::{ControllerEvent, Parser, ProtocolParseError};
+///
+/// /// internal test helper
+/// fn assert_is_button_event(
+///     event: &Result<ControllerEvent, ProtocolParseError>,
+///     button: Button,
+///     button_state: ButtonState,
+/// ) {
+///     match event {
+///         Ok(ButtonEvent(event)) => {
+///             assert_eq!(event.button(), &button);
+///             assert_eq!(event.state(), &button_state)
+///         }
+///         _ => assert!(false),
+///     }
+/// }
+///
+/// // the example input contains some null bytes, two button events and two malformed events.
+/// let input = b"\x00!B11:!B10;\x00\x00!\x00\x00\x00\x00!B138";
+/// let mut parser = Parser::new(input);
+///
+/// assert_is_button_event(
+///     &parser.next().unwrap(),
+///     Button::Button1,
+///     ButtonState::Pressed,
+/// );
+/// assert_is_button_event(
+///     &parser.next().unwrap(),
+///     Button::Button1,
+///     ButtonState::Released,
+/// );
+/// assert_eq!(
+///     parser.next().unwrap(),
+///     Err(ProtocolParseError::UnknownEvent(Some(0)))
+/// );
+/// if let Err(e) = &parser.next().unwrap() {
+///     assert_eq!(
+///         e,
+///         &ProtocolParseError::ButtonParseError(ButtonParseError::UnknownButtonState(b'3'))
+///     );
+///     # {
+///         // test the `core::error::Error` implementation
+///         # extern crate alloc;
+///         # use alloc::string::ToString;
+///         # use core::error::Error;
+///         assert_eq!(
+///             e.source().unwrap().to_string(),
+///             "Unknown button state: 0x33"
+///         );
+///     # }
+/// } else {
+///     assert!(false, "expected an error");
+/// }
+/// assert_eq!(parser.next(), None);
+/// ```
+#[derive(Debug, Copy, Clone)]
+pub struct Parser<'a> {
+    input: &'a [u8],
+    curr_pos: usize,
+}
+
+impl<'a> Parser<'a> {
+    /// Create a new parser. The input is parsed step by step on each invocation of `next`.
+    pub fn new(input: &'a [u8]) -> Self {
+        Self { input, curr_pos: 0 }
     }
-    let mut state = ParserState::SeekStart;
+}
 
-    let mut result = Vec::new();
+impl<'a> Iterator for Parser<'a> {
```

Comment:
This causes a clippy warning on nightly:

```
    Checking adafruit-bluefruit-protocol v0.2.0 (/tmp/adafruit-bluefruit-protocol-rs)
error: the following explicit lifetimes could be elided: 'a
   --> src/lib.rs:278:6
    |
278 | impl<'a> Iterator for Parser<'a> {
    |      ^^                      ^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
note: the lint level is defined here
   --> src/lib.rs:17:9
    |
17  | #![deny(warnings)]
    |         ^^^^^^^^
    = note: `#[deny(clippy::needless_lifetimes)]` implied by `#[deny(warnings)]`
help: elide the lifetimes
    |
278 - impl<'a> Iterator for Parser<'a> {
278 + impl Iterator for Parser<'_> {
    |

error: could not compile `adafruit-bluefruit-protocol` (lib) due to 1 previous error
```

Following the advice leads to code that builds without warnings on both stable and nightly.
```suggestion
impl Iterator for Parser<'_> {
```

Reply:
thanks for testing with nightly! fixed